libsolidity/parsing/Parser.cpp

```diff
@@ -329,15 +332,33 @@ StateMutability Parser::parseStateMutability(Token::Value _token)
 	return stateMutability;
 }
 
-Parser::FunctionHeaderParserResult Parser::parseFunctionHeader(bool _forceEmptyName, bool _allowModifiers)
+Parser::FunctionHeaderParserResult Parser::parseFunctionHeader(
+	bool _forceEmptyName,
+	bool _allowModifiers,
+	ASTString const* _contractName
+)
 {
 	RecursionGuard recursionGuard(*this);
 	FunctionHeaderParserResult result;
-	expectToken(Token::Function);
-	if (_forceEmptyName || m_scanner->currentToken() == Token::LParen)
-		result.name = make_shared<ASTString>(); // anonymous function
+
+	if (m_scanner->currentToken() == Token::Function)
+		// In case of old style constructors, i.e. functions with the same name as the contract,
```

Comment:
It might be easier to understand if you just set it to false unconditionally before this if statement. Everyone would exect it to be reset to true at some later point in this function.